src/helpers/convert.rs

```diff
@@ -0,0 +1,49 @@
+use implementations::NarrowerThan;
+
+/// Replacement for `as` casts going from wide integer to narrower integer.
+///
+/// # Example
+///
+/// ```ignore
+/// let x = 99_u64;
+/// let lo = x.truncate::<u16>();
+/// // lo is of type u16, equivalent to `x as u16`.
+/// ```
+pub(crate) trait Truncate: Sized {
+    fn truncate<To>(self) -> To
+    where
+        To: NarrowerThan<Self>,
+    {
+        NarrowerThan::truncate_from(self)
+    }
+}
+
+impl Truncate for u16 {}
+impl Truncate for u32 {}
+impl Truncate for u64 {}
+impl Truncate for u128 {}
+
+mod implementations {
+    pub(crate) trait NarrowerThan<T> {
+        fn truncate_from(wide: T) -> Self;
+    }
+
+    macro_rules! impl_narrower_than {
+        ($(NarrowerThan<{$($ty:ty),*}> for $self:ty)*) => {
+            $($(
+                impl NarrowerThan<$ty> for $self {
+                    fn truncate_from(wide: $ty) -> Self {
```

Comment:
Do you think these should be `inline(always)` just to be sure?

Reply:
Even if this code were in my most performance-sensitive crates, I would not expect to need `inline(always)` here. In the standard library even `From<T> for T` and `From<u8> for u16` do not have `inline(always)`.